src/humanloop/otel/processor.py

```diff
@@ -0,0 +1,194 @@
+        return
+    else:
+        logging.error("Invalid span type")
+
+
+def _process_prompt(prompt_span: ReadableSpan, children_spans: list[ReadableSpan]):
+    if len(children_spans) == 0:
+        return
+    child_span = children_spans[0]
+    assert is_llm_provider_call(child_span)
+    _enrich_prompt_span_file(prompt_span, child_span)
+    _enrich_prompt_span_log(prompt_span, child_span)
+
+
+def _process_tool(tool_span: ReadableSpan, children_spans: list[ReadableSpan]):
+    tool_log = read_from_opentelemetry_span(tool_span, key=HL_LOG_OT_KEY)
+    if tool_span.start_time:
+        tool_log["start_time"] = tool_span.start_time / 1e9
+    if tool_span.end_time:
+        tool_log["end_time"] = tool_span.end_time / 1e9
+        tool_log["created_at"] = tool_span.end_time / 1e9
+
+    write_to_opentelemetry_span(
+        span=tool_span,
+        key=HL_LOG_OT_KEY,
+        value=tool_log,
+    )
+
+
+def _process_flow(flow_span: ReadableSpan, children_spans: list[ReadableSpan]):
+    # TODO: Use children_spans in the future
+    flow_log = read_from_opentelemetry_span(flow_span, key=HL_LOG_OT_KEY)
+    if flow_span.start_time:
+        flow_log["start_time"] = flow_span.start_time / 1e9
+    if flow_span.end_time:
+        flow_log["end_time"] = flow_span.end_time / 1e9
+        flow_log["created_at"] = flow_span.end_time / 1e9
+
+    write_to_opentelemetry_span(
+        span=flow_span,
+        key=HL_LOG_OT_KEY,
+        value=flow_log,
+    )
+
+
+def _enrich_prompt_span_file(prompt_span: ReadableSpan, llm_provider_call_span: ReadableSpan):
+    hl_file: dict[str, Any] = read_from_opentelemetry_span(prompt_span, key=HL_FILE_OT_KEY)
+    gen_ai_object: dict[str, Any] = read_from_opentelemetry_span(llm_provider_call_span, key="gen_ai")
+    llm_object: dict[str, Any] = read_from_opentelemetry_span(llm_provider_call_span, key="llm")
+
+    prompt = hl_file.get("prompt")
+    if not prompt:
+        prompt = {}
+    if not prompt.get("model"):
+        prompt["model"] = gen_ai_object.get("request", {}).get("model", None)
+    if not prompt.get("endpoint"):
+        prompt["endpoint"] = llm_object.get("request", {}).get("type")
+    if not prompt.get("provider"):
+        prompt["provider"] = gen_ai_object.get("system", None)
+        if prompt["provider"]:
+            prompt["provider"] = prompt["provider"].lower()
+    if not prompt.get("temperature"):
+        prompt["temperature"] = gen_ai_object.get("request", {}).get("temperature", None)
+    if not prompt.get("top_p"):
+        prompt["top_p"] = gen_ai_object.get("request", {}).get("top_p", None)
+    if not prompt.get("max_tokens"):
+        prompt["max_tokens"] = gen_ai_object.get("request", {}).get("max_tokens", None)
+    if not prompt.get("presence_penalty"):
+        prompt["presence_penalty"] = llm_object.get("presence_penalty", None)
+    if not prompt.get("frequency_penalty"):
+        prompt["frequency_penalty"] = llm_object.get("frequency_penalty", None)
```

Comment:
Could you confirm that the param names are in fact common across all providers?

Reply:
Yes, see comment above https://github.com/humanloop/humanloop-python/pull/24#discussion_r1824111431